### PR TITLE
Fix unstable sort order in `ImportExport::export()` might lead to duplicate/missing entries

### PR DESCRIPTION
**Pull request by Nevay**
### 1. Why is this change necessary?
`ImportExport::export()` does not guarantee a stable order for the entities. Entities around `$exportLimit` might be exported a second time / skipped on the next call to `::export()` if they compare as equal according to the provided sorting.

### 2. What does this change do, exactly?
Adds a tiebreaker sorting to ensure a consistent order.

### 3. Describe each step to reproduce the issue or behaviour.
`ImportExportTest::testDefaultProperties()` can trigger this issue.

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Content/ImportExport/ImportExport.php [hunk 1]
```diff
@@ -297,6 +297,7 @@ public function export(Context $context, ?Criteria $criteria = null, int $offset
             // default sorting
             $criteria->addSorting(new FieldSorting('createdAt', FieldSorting::ASCENDING));
         }
+        $criteria->addSorting(new FieldSorting('id'));
```
Comment (Nevay) [inline src/Core/Content/ImportExport/ImportExport.php:300]: Should this use something like the following instead of relying on a hardcoded `id` field (was based on the hardcoded `createdAt` field above)?
```suggestion
        foreach ($this->repository->getDefinition()->getPrimaryKeys() as $primaryKey) {
            $criteria->addSorting(new FieldSorting($primaryKey->getPropertyName()));
        }
```

Comment (dfrancos-hub) [reply src/Core/Content/ImportExport/ImportExport.php:300]: Hi, it's not necessary in this case. We can proceed with the PR. Thanks

**Comment by keulinho** [comment]
Your changes were merged, thanks again for your contribution! :blue_heart:

**Comment by macster84** [comment]
Hi @Nevay ,
unfortunately the introduced change breaks the export of join tables for n:m relations. These do not have an 'id' field but rather two ids pointing to the respective tables of the relation. They make up a compound primary key.

In my case this table has an additional column besides just the ids. Maybe there is a better/canonical way to export such tables. What do you think?

Currently I am inclined to open another PR to address that issue.

**Comment by Nevay** [comment]
Hi @macster84, can you test whether https://github.com/shopware/platform/pull/2658#discussion_r948979144 would solve the issue? If yes, feel free to open a PR with that change.

**Comment by macster84** [comment]
Hi @Nevay ,

thanks for the quick response. I can confirm that my manual test worked with your proposal.
I will open a PR.

**Comment by macster84** [comment]
Hi again @Nevay,


I dived into existing Shopware code and documentation.
**TL;DR:**
The current solution created by this PR is probably good enough to export enhanced many-to-many relations. So those ones where the mapping table has additional columns. 
It remains unclear to me how pure mapping tables should be exported.


### More detailed response:

I think we can differentiate many-to-many export into two questions:
1) pure many-to-many mapping entity
2) mapping entity with additional fields

Pure many-to-many mapping entity
---

Pure many-to-many entities extend MappingEntityDefinition. There are two foreign key fields pointing to each side of the relation. These fields also make up the primary key. That's it.
AFAIK these kind of entities cannot be exported via Shopware for two reasons at least:
a) They do not have a _createdAt_ field
b) They do not have an _id_ field. 

While the second reason was introduced by this PR, the first one was already there. I noticed it when I tried to write a test to export such a mapping entity _ProductCategoryTree_.
I am truly unsure how pure many-to-many entities shall be exported with Shopware at all. I could not find anything via Google and will probably ask this question on Shopware's Slack chat.

Mapping entity with additional fields
---

The other side of the story are mapping entities with additional fields. They are also somehow covered in the [developer documentation](https://developer.shopware.com/docs/guides/plugins/plugins/framework/data-handling/deleting-associated-data#deleting-onetomany-associations):

>The OneToMany associations deserve an own section, since they come with a special use case in Shopware 6. They are sometimes used to create a ManyToMany association but with extra data in the mapping table, and sometimes they're just simple OneToMany associations.
>
>[...]
>
>Unfortunately, it's not always that simple. As explained above, sometimes OneToMany associations are hidden ManyToMany associations. To understand what we mean, have a look at the media field in the [Product definition](https://github.com/shopware/platform/blob/v6.3.4.0/src/Core/Content/Product/ProductDefinition.php#L210-L211). Technically a product can have multiple medias, and a media can be assigned to multiple products, so this should have been a ManyToMany association, right? Yet, looking at the media field in the ProductDefinition, you can see that it's a OneToMany association. The second case, that we described earlier in this section, fits here: Technically a ManyToMany association, hidden by a OneToMany association for the reason mentioned above: There's more data needed for the mapping entity.


The [ProductMediaDefinition](https://github.com/shopware/platform/blob/v6.4.18.1/src/Core/Content/Product/Aggregate/ProductMedia/ProductMediaDefinition.php) actually carries an additional _id_ field even though it is used for representing a many-to-many relation. Another example is the [MailTemplateMediaDefinition](https://github.com/shopware/platform/blob/v6.4.18.1/src/Core/Content/MailTemplate/Aggregate/MailTemplateMedia/MailTemplateMediaDefinition.php). This one has less additional fields and more resembles the use case I am having. But even that one carries another _id_ field. Interestingly the _ProductMediaDefinition_ has a _createdAt_  field while the _MailTemplateMediaDefinition_ has not. There seems to be some inconsistency here.

I guess you can argue that the existence of additional fields qualifies the mapping entity as an entity on its own. 
And an entity should have an _id_. And for the same reason a _createdAt_ field. 

So long story short
---
Export of pure many-to-many mapping entities would not work even with your proposal. We would still need a _createdAt_ column in the table.
For enhanced many-to-many mapping entities, the Shopware code seems to indicate a best practice of having an _id_ and a _createdAt_ field. For the latter the export already works today.
I guess I will ask a question to the Shopware Slack channel before proceeding any further.
